src/Workspaces/Core/Portable/CodeActions/CodeAction_Cleanup.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CaseCorrection;
+using Microsoft.CodeAnalysis.CodeCleanup;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Simplification;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeActions;
+
+public abstract partial class CodeAction
+{
+    /// <summary>
+    /// We do cleanup in N serialized passes.  This allows us to process all documents in parallel, while only forking
+    /// the solution N times *total* (instead of N times *per* document).
+    /// </summary>
+    private static readonly ImmutableArray<Func<Document, CodeCleanupOptions, CancellationToken, Task<Document>>> s_cleanupPasses =
+    [
+        // First, ensure that everything is formatted as the feature asked for.  We want to do this prior to doing
+        // semantic cleanup as the semantic cleanup passes may end up making changes that end up dropping some of
+        // the formatting/elastic annotations that the feature wanted respected.
+        static (document, options, cancellationToken) => CleanupSyntaxAsync(document, options, cancellationToken),
+        // Then add all missing imports to all changed documents.
+        static (document, options, cancellationToken) => ImportAdder.AddImportsFromSymbolAnnotationAsync(document, Simplifier.AddImportsAnnotation, options.AddImportOptions, cancellationToken),
+        // Then simplify any expanded constructs.
+        static (document, options, cancellationToken) => Simplifier.ReduceAsync(document, Simplifier.Annotation, options.SimplifierOptions, cancellationToken),
+        // The do any necessary case correction for VB files.
+        static (document, options, cancellationToken) => CaseCorrector.CaseCorrectAsync(document, CaseCorrector.Annotation, cancellationToken),
+        // Finally, after doing the semantic cleanup, do another syntax cleanup pass to ensure that the tree is in a
+        // good state. The semantic cleanup passes may have introduced new nodes with elastic trivia that have to be
+        // cleaned.
+        static (document, options, cancellationToken) => CleanupSyntaxAsync(document, options, cancellationToken),
+    ];
```

Comment:
i really liked this.  it emphasizes that we're doing a series of passes in a specific order.